### ISSUE TITLE
linux: `abort()` on EBADF/ENOENT from epoll_ctl(ADD) races with concurrent close()

### ISSUE THREAD

**Issue by ehthayer** [libuv/libuv#5130]
## Problem

On Linux, `uv__epoll_ctl_flush` and `uv__io_poll` call `abort()` when `epoll_ctl(ADD)` returns an unexpected errno — including `EBADF` / `ENOENT` from a file descriptor that was closed from another thread between libuv staging the op and the kernel executing it. The result is a process-wide SIGABRT triggered by a timing race, not an internal libuv error.

## Context

Inside the same loop, `EPOLL_CTL_DEL` already tolerates this exact race. From `uv__epoll_ctl_prep`:

```c
if (op == EPOLL_CTL_DEL)
  continue;  /* Ignore errors, may be racing with another thread. */
```

`EPOLL_CTL_ADD` / `EPOLL_CTL_MOD` currently don't apply that same tolerance. If a caller races a `close()` against libuv's next `epoll_ctl` on a registered fd, the abort fires.

This is straightforward to reproduce deterministically: register an fd with `uv_poll_start`, `close(2)` it directly (bypassing `uv_close`), call `uv_poll_start` again to change the event mask, run the loop. The kernel returns `EBADF`; libuv aborts.

See #5129 for a reproducer test and proposed minimal patch.

## Observed in production

uvloop 0.22.1 (vendored libuv 1.48.0) running a Python/asyncio HTTP server with many concurrent WebSocket connections. Core dumps show SIGABRTs from both the io_uring CQE path (`cqe->res != -EEXIST abort()`) and the classic `uv__io_poll` path (`assert(errno == EEXIST)`). Rate: ~one SIGABRT every 5–10 minutes under ~30 WebSockets/sec churn on a 2-CPU pod. A related user-facing bug report on the Python side is [MagicStack/uvloop#738](https://github.com/MagicStack/uvloop/issues/738).

## Question for maintainers

Is the right place for this tolerance libuv (applying the existing CTL_DEL rule uniformly to CTL_ADD) or the caller (e.g. uvloop tightening fd ownership)?

I lean libuv — any wrapper that hands libuv a borrowed fd hits this, not just uvloop, and CTL_DEL already accepted that premise. But happy to discuss here rather than on the PR.

**Comment by bnoordhuis** [comment]
Closing because things are working as intended.

It looks like you used an LLM to summarize linux.c? uv__epoll_ctl_prep does not have any lines or comments resembling what you posted.

**Comment by ehthayer** [comment]
Brief follow-up: spent the last couple of days trying to fix this race in user-space (in `aiohappyeyeballs`, the Happy Eyeballs implementation that aiohttp uses internally), since the closing comment on this issue and on #5129 implied the right place is the caller.

Three fix variants, all verified live, all still produced SIGABRT under a reliable reproducer:

1. Selector-aware close (`loop.remove_writer/reader(fd)` before `sock.close()`)
2. Deferred close (`sock.detach()` + `loop.call_soon(os.close, fd)`)
3. Both above + patching the runner-up cleanup in `aiohappyeyeballs.start_connection`

Posting the full evidence + analysis on the PR thread: #5129 (comment).

Short version: the race fires from socket-close paths beyond `aiohappyeyeballs`' reach (aiohttp's WebSocket transport, the `websockets` library, etc.), all of which interact with libuv's lazy `epoll_ctl` staging the same way. Patching one Python library doesn't help because the others still race. Whack-a-mole that we don't think is winnable from user-space.

The structural argument is unchanged from #5129: `uv__epoll_ctl_prep` already tolerates the same race for `EPOLL_CTL_DEL` (`/* Ignore errors, may be racing with another thread. */`). Extending the same tolerance to `ADD`/`MOD` for `EBADF`/`ENOENT` is the structurally correct fix at libuv's level — what we did empirically supports that conclusion now in a way it didn't when this was first filed.
